File: LoRDeckCodes_py3/lor_deck_encoder.py
from base64 import b32encode, b32decode

from card_code_and_count import CardCodeAndCount
import varint_translator

CARD_CODE_LENGTH = 7
faction_to_int = {"DE": 0, "FR": 1, "IO": 2, "NX": 3, "PZ": 4, "SI": 5}
int_to_faction = {0: "DE", 1: "FR", 2: "IO", 3: "NX", 4: "PZ", 5: "SI"}
MAX_KNOWN_VERSION = 1


def get_deck_from_code(code):
    result = []

    try:
        while len(code) % 8:
            code += "="
        bytes_ = b32decode(code)
    except Exception:
        raise ValueError("Invalid deck code")

    byte_list = bytearray(bytes_)

    # grab format and version
    format_ = bytes_[0] >> 4
    version = bytes_[0] & 15
    byte_list.pop(0)

    if version > MAX_KNOWN_VERSION:
        raise ValueError("The provided code requires a higher version \
                         of this library; please update.")

    for i in range(3, 0, -1):
        byte_list, num_group_ofs = varint_translator.pop_varint(byte_list)

        for j in range(num_group_ofs):
            byte_list, num_ofs_in_this_group = varint_translator.pop_varint(
                                                                     byte_list)
            byte_list, set_ = varint_translator.pop_varint(byte_list)
            byte_list, faction = varint_translator.pop_varint(byte_list)

            for k in range(num_ofs_in_this_group):
                byte_list, card = varint_translator.pop_varint(byte_list)

                set_string = str(set_).zfill(2)
                faction_string = int_to_faction[faction]
                card_string = str(card).zfill(3)

                new_entry = CardCodeAndCount(set_string + faction_string
                                             + card_string, i)
                result.append(new_entry)

    # the remainder of the deck code is comprised
    # of entries for cards with counts >= 4.
    # this will only happen in Limited and special game modes.
    # the encoding is simply [count] [cardcode]
    while byte_list:
        byte_list, four_plus_count = varint_translator.pop_varint(byte_list)
        byte_list, four_plus_set = varint_translator.pop_varint(byte_list)
        byte_list, four_plus_faction = varint_translator.pop_varint(byte_list)
        byte_list, four_plus_number = varint_translator.pop_varint(byte_list)

        four_plus_set_string = str(four_plus_set).zfill(2)
        four_plus_faction_string = int_to_faction[four_plus_faction]
        four_plus_number_string = str(four_plus_number).zfill(3)

        new_entry = CardCodeAndCount(four_plus_set_string
                                     + four_plus_faction_string
                                     + four_plus_number_string,
                                     four_plus_count)
        result.append(new_entry)

    return result


def get_code_from_deck(deck):
    result = b32encode(get_deck_code_bytes(deck)).replace(b'=', b'')
    return result.decode()  # return a string and not a byte string


def get_deck_code_bytes(deck):
    result = bytearray()

    if not valid_card_codes_and_count(deck):
        raise ValueError("The provided deck contains invalid card codes.")

    format_and_version = bytes([17])  # i.e. 00010001
    result.extend(format_and_version)

    of_3 = []
    of_2 = []
    of_1 = []
    of_n = []

    for ccc in deck:
        if ccc.count == 3:
            of_3.append(ccc)
        elif ccc.count == 2:
            of_2.append(ccc)
        elif ccc.count == 1:
            of_1.append(ccc)
        elif ccc.count < 1:
            raise ValueError("Invalid count of " + ccc.count +
                             " for card " + ccc.card_code)
        else:
            of_n.append(ccc)

    # build the lists of set and faction combinations
    # within the groups of similar card counts
    grouped_of_3s = get_grouped_ofs(of_3)
    grouped_of_2s = get_grouped_ofs(of_2)
    grouped_of_1s = get_grouped_ofs(of_1)

    # to ensure that the same decklist in any order
    # produces the same code, do some sorting
    grouped_of_3s = sort_group_of(grouped_of_3s)
    grouped_of_2s = sort_group_of(grouped_of_2s)
    grouped_of_1s = sort_group_of(grouped_of_1s)

    # Nofs (since rare) are simply sorted by the card code
    # there's no optimiziation based upon the card count
    of_n = sorted(of_n, key=lambda c: c.card_code)

    # Encode
    encode_group_of(result, grouped_of_3s)
    encode_group_of(result, grouped_of_2s)
    encode_group_of(result, grouped_of_1s)

    # Cards with 4+ counts are handled differently:
    # simply [count] [card code] for each
    encode_n_ofs(result, of_n)

    return result


def encode_n_ofs(bytes_, n_ofs):
    for ccc in n_ofs:
        bytes_.extend(varint_translator.get_varint(ccc.count))

        set_number, faction_code, card_number = parse_card_code(ccc.card_code)
        faction_number = faction_to_int[faction_code]

        bytes_.extend(varint_translator.get_varint(set_number))
        bytes_.extend(varint_translator.get_varint(faction_number))
        bytes_.extend(varint_translator.get_varint(card_number))


# The sorting convention of this encoding scheme is
# First by the number of set/faction combinations in each top-level list
# Second by the alphanumeric order of the card codes within those lists.
def sort_group_of(group_of):
    group_of = sorted(group_of, key=lambda l: len(l))

    for i in range(len(group_of)):
        group_of[i] = sorted(group_of[i], key=lambda c: c.card_code)

    return group_of


def parse_card_code(code):
    set_, faction, number = int(code[0:2]), code[2:4], int(code[4:7])
    return set_, faction, number


def get_grouped_ofs(list_):
    result = []
    while list_:
        current_set = []

        # get info from first
        first_card_code = list_[0].card_code
        set_number, faction_code, card_number = parse_card_code(
                                                               first_card_code)

        # now add that to our new list, remove from old
        current_set.append(list_[0])
        list_.pop(0)

        # sweep through rest of list
        # and grab entries that should live with our first one.
        # matching means same set and faction
        # we are already assured the count matches from previous grouping.
        for i in range(len(list_)):
            current_card_code = list_[i].card_code
            current_set_number = int(current_card_code[0:2])
            current_faction_code = current_card_code[2:4]

            if current_set_number == set_number \
               and current_faction_code == faction_code:
                current_set.append(list_[i])

        for card in current_set[1:]:
            list_.remove(card)

        result.append(current_set)
    return result


def encode_group_of(bytes_, group_of):
    bytes_.extend(varint_translator.get_varint(len(group_of)))

    for current_list in group_of:
        # how many cards in current group?
        bytes_.extend(varint_translator.get_varint(len(current_list)))

        # what is this group, as identified by a set and faction pair
        current_card_code = current_list[0].card_code

        current_set_number, current_faction_code, _ = parse_card_code(
                                                             current_card_code)
        current_faction_number = faction_to_int[current_faction_code]

        bytes_.extend(varint_translator.get_varint(current_set_number))
        bytes_.extend(varint_translator.get_varint(current_faction_number))

        # what are the cards, as identified by the third section of card code
        # only now, within this group?
        for cd in current_list:
            code = cd.card_code
            sequence_number = int(code[4:7])
            bytes_.extend(varint_translator.get_varint(sequence_number))


def valid_card_codes_and_count(deck):
    for ccc in deck:
        if len(ccc.card_code) != CARD_CODE_LENGTH:
            return False

        try:
            int(ccc.card_code[0:2])
        except ValueError:
            return False

        if ccc.card_code[2:4] not in faction_to_int.keys():
            return False

        try:
            int(ccc.card_code[4:7])
        except ValueError:
            return False

        if ccc.count < 1:
            return False

    return True

File: LoRDeckCodes/LoRDeckEncoder.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace LoRDeckCodes
{
    public class LoRDeckEncoder
    {
        private readonly static int CARD_CODE_LENGTH = 7;
        private static Dictionary<string, int> FactionCodeToIntIdentifier = new Dictionary<string, int>();
        private static Dictionary<int, string> IntIdentifierToFactionCode = new Dictionary<int, string>();
        private static readonly Dictionary<string, int> FactionCodeToLibraryVersion = new Dictionary<string, int>();
        private readonly static int MAX_KNOWN_VERSION = 4;
        private readonly static int FORMAT = 1;
        private readonly static int INITIAL_VERSION = 1;


        static LoRDeckEncoder()
        {
            PopulateFactionCodes();
            PopulateFactionLibraryVersions();
        }

        private static void PopulateFactionCodes()
        {
            FactionCodeToIntIdentifier.Add("DE", 0);
            FactionCodeToIntIdentifier.Add("FR", 1);
            FactionCodeToIntIdentifier.Add("IO", 2);
            FactionCodeToIntIdentifier.Add("NX", 3);
            FactionCodeToIntIdentifier.Add("PZ", 4);
            FactionCodeToIntIdentifier.Add("SI", 5);
            FactionCodeToIntIdentifier.Add("BW", 6);
            FactionCodeToIntIdentifier.Add("SH", 7);
            FactionCodeToIntIdentifier.Add("MT", 9);
            FactionCodeToIntIdentifier.Add("BC", 10);
            IntIdentifierToFactionCode.Add(0, "DE");
            IntIdentifierToFactionCode.Add(1, "FR");
            IntIdentifierToFactionCode.Add(2, "IO");
            IntIdentifierToFactionCode.Add(3, "NX");
            IntIdentifierToFactionCode.Add(4, "PZ");
            IntIdentifierToFactionCode.Add(5, "SI");
            IntIdentifierToFactionCode.Add(6, "BW");
            IntIdentifierToFactionCode.Add(7, "SH"); 
            IntIdentifierToFactionCode.Add(9, "MT"); 
            IntIdentifierToFactionCode.Add(10, "BC");
        }
        private static void PopulateFactionLibraryVersions()
        {
            FactionCodeToLibraryVersion.Add("DE", 1);
            FactionCodeToLibraryVersion.Add("FR", 1);
            FactionCodeToLibraryVersion.Add("IO", 1);
            FactionCodeToLibraryVersion.Add("NX", 1);
            FactionCodeToLibraryVersion.Add("PZ", 1);
            FactionCodeToLibraryVersion.Add("SI", 1);
            FactionCodeToLibraryVersion.Add("BW", 2);
            FactionCodeToLibraryVersion.Add("MT", 2);
            FactionCodeToLibraryVersion.Add("SH", 3);
            FactionCodeToLibraryVersion.Add("BC", 4);
        }

        public static List<CardCodeAndCount> GetDeckFromCode(string code)
        {
            List<CardCodeAndCount> result = new List<CardCodeAndCount>();

            byte[] bytes = null;
            try
            {
                bytes = Base32.Decode(code);
            }
            catch
            {
                throw new ArgumentException("Invalid deck code");
            }

            List<byte> byteList = bytes.ToList();

            //grab format and version
            int format = bytes[0] >> 4;
            int version = bytes[0] & 0xF;
            byteList.RemoveAt(0);

            if (version > MAX_KNOWN_VERSION)
            {
                throw new ArgumentException("The provided code requires a higher version of this library; please update.");
            }

            for (int i = 3; i > 0; i--)
            {
                int numGroupOfs = VarintTranslator.PopVarint(byteList);

                for (int j = 0; j < numGroupOfs; j++)
                {
                    int numOfsInThisGroup = VarintTranslator.PopVarint(byteList);
                    int set = VarintTranslator.PopVarint(byteList);
                    int faction = VarintTranslator.PopVarint(byteList);

                    for (int k = 0; k < numOfsInThisGroup; k++)
                    {
                        int card = VarintTranslator.PopVarint(byteList);

                        string setString = set.ToString().PadLeft(2, '0');
                        string factionString = IntIdentifierToFactionCode[faction];
                        string cardString = card.ToString().PadLeft(3, '0');

                        CardCodeAndCount newEntry = new CardCodeAndCount() { CardCode = setString + factionString + cardString, Count = i };
                        result.Add(newEntry);
                    }

                }
            }

            //the remainder of the deck code is comprised of entries for cards with counts >= 4
            //this will only happen in Limited and special game modes.
            //the encoding is simply [count] [cardcode]
            while (byteList.Count > 0)
            {
                int fourPlusCount = VarintTranslator.PopVarint(byteList);
                int fourPlusSet = VarintTranslator.PopVarint(byteList);
                int fourPlusFaction = VarintTranslator.PopVarint(byteList);
                int fourPlusNumber = VarintTranslator.PopVarint(byteList);

                string fourPlusSetString = fourPlusSet.ToString().PadLeft(2, '0');
                string fourPlusFactionString = IntIdentifierToFactionCode[fourPlusFaction];
                string fourPlusNumberString = fourPlusNumber.ToString().PadLeft(3, '0');

                CardCodeAndCount newEntry = new CardCodeAndCount() { CardCode = fourPlusSetString + fourPlusFactionString + fourPlusNumberString, Count = fourPlusCount };
                result.Add(newEntry);
            }

            return result;
        }

        public static string GetCodeFromDeck(List<CardCodeAndCount> deck)
        {
            string result = Base32.Encode(GetDeckCodeBytes(deck));
            return result;
        }

        private static byte[] GetDeckCodeBytes(List<CardCodeAndCount> deck)
        {
            List<byte> result = new List<byte>();

            if (!ValidCardCodesAndCounts(deck))
                throw new ArgumentException("The provided deck contains invalid card codes.");

            byte formatAndVersion = (byte)(FORMAT << 4 | (GetMinSupportedLibraryVersion(deck) & 0xF));

            result.Add(formatAndVersion);

            List<CardCodeAndCount> of3 = new List<CardCodeAndCount>();
            List<CardCodeAndCount> of2 = new List<CardCodeAndCount>();
            List<CardCodeAndCount> of1 = new List<CardCodeAndCount>();
            List<CardCodeAndCount> ofN = new List<CardCodeAndCount>();

            foreach (CardCodeAndCount ccc in deck)
            {
                if (ccc.Count == 3)
                    of3.Add(ccc);
                else if (ccc.Count == 2)
                    of2.Add(ccc);
                else if (ccc.Count == 1)
                    of1.Add(ccc);
                else if (ccc.Count < 1)
                    throw new ArgumentException("Invalid count of " + ccc.Count + " for card " + ccc.CardCode);
                else
                    ofN.Add(ccc);
            }

            //build the lists of set and faction combinations within the groups of similar card counts
            List<List<CardCodeAndCount>> groupedOf3s = GetGroupedOfs(of3);
            List<List<CardCodeAndCount>> groupedOf2s = GetGroupedOfs(of2);
            List<List<CardCodeAndCount>> groupedOf1s = GetGroupedOfs(of1);

            //to ensure that the same decklist in any order produces the same code, do some sorting
            groupedOf3s = SortGroupOf(groupedOf3s);
            groupedOf2s = SortGroupOf(groupedOf2s);
            groupedOf1s = SortGroupOf(groupedOf1s);

            //Nofs (since rare) are simply sorted by the card code - there's no optimiziation based upon the card count
            ofN = ofN.OrderBy(c => c.CardCode).ToList();

            //Encode
            EncodeGroupOf(result, groupedOf3s);
            EncodeGroupOf(result, groupedOf2s);
            EncodeGroupOf(result, groupedOf1s);

            //Cards with 4+ counts are handled differently: simply [count] [card code] for each
            EncodeNOfs(result, ofN);

            return result.ToArray();
        }
        private static int GetMinSupportedLibraryVersion(List<CardCodeAndCount> deck)
        {
            if (!deck.Any())
            {
                return INITIAL_VERSION;
            }

            return deck
                .Select(ccc => ccc.CardCode.Substring(2,2))
                .Select(factionCode => FactionCodeToLibraryVersion.TryGetValue(factionCode, out var version) ? version : MAX_KNOWN_VERSION)
                .Max();
        }

        private static void EncodeNOfs(List<byte> bytes, List<CardCodeAndCount> nOfs)
        {
            foreach (CardCodeAndCount ccc in nOfs)
            {
                bytes.AddRange(VarintTranslator.GetVarint(ccc.Count));

                ParseCardCode(ccc.CardCode, out int setNumber, out string factionCode, out int cardNumber);
                int factionNumber = FactionCodeToIntIdentifier[factionCode];

                bytes.AddRange(VarintTranslator.GetVarint(setNumber));
                bytes.AddRange(VarintTranslator.GetVarint(factionNumber));
                bytes.AddRange(VarintTranslator.GetVarint(cardNumber));
            }
        }

        //The sorting convention of this encoding scheme is
        //First by the number of set/faction combinations in each top-level list
        //Second by the alphanumeric order of the card codes within those lists.
        private static List<List<CardCodeAndCount>> SortGroupOf(List<List<CardCodeAndCount>> groupOf)
        {
            groupOf = groupOf.OrderBy(g => g.Count).ThenBy(c => c[0].CardCode).ToList();
            for (int i = 0; i < groupOf.Count; i++)
            {
                groupOf[i] = groupOf[i].OrderBy(c => c.CardCode).ToList();
            }
            return groupOf;
        }

        private static void ParseCardCode(string code, out int set, out string faction, out int number)
        {
            set = int.Parse(code.Substring(0, 2));
            faction = code.Substring(2, 2);
            number = int.Parse(code.Substring(4, 3));
        }

        private static List<List<CardCodeAndCount>> GetGroupedOfs(List<CardCodeAndCount> list)
        {
            List<List<CardCodeAndCount>> result = new List<List<CardCodeAndCount>>();
            while (list.Count > 0)
            {
                List<CardCodeAndCount> currentSet = new List<CardCodeAndCount>();

                //get info from first
                string firstCardCode = list[0].CardCode;
                ParseCardCode(firstCardCode, out int setNumber, out string factionCode, out int cardNumber);

                //now add that to our new list, remove from old
                currentSet.Add(list[0]);
                list.RemoveAt(0);

                //sweep through rest of list and grab entries that should live with our first one.
                //matching means same set and faction - we are already assured the count matches from previous grouping.
                for (int i = list.Count - 1; i >= 0; i--)
                {
                    string currentCardCode = list[i].CardCode;
                    int currentSetNumber = int.Parse(currentCardCode.Substring(0, 2));
                    string currentFactionCode = currentCardCode.Substring(2, 2);

                    if (currentSetNumber == setNumber && currentFactionCode == factionCode)
                    {
                        currentSet.Add(list[i]);
                        list.RemoveAt(i);
                    }
                }
                result.Add(currentSet);
            }
            return result;
        }

        private static void EncodeGroupOf(List<byte> bytes, List<List<CardCodeAndCount>> groupOf)
        {
            bytes.AddRange(VarintTranslator.GetVarint(groupOf.Count));
            foreach (List<CardCodeAndCount> currentList in groupOf)
            {
                //how many cards in current group?
                bytes.AddRange(VarintTranslator.GetVarint(currentList.Count));

                //what is this group, as identified by a set and faction pair
                string currentCardCode = currentList[0].CardCode;
                ParseCardCode(currentCardCode, out int currentSetNumber, out string currentFactionCode, out int _);
                int currentFactionNumber = FactionCodeToIntIdentifier[currentFactionCode];
                bytes.AddRange(VarintTranslator.GetVarint(currentSetNumber));
                bytes.AddRange(VarintTranslator.GetVarint(currentFactionNumber));

                //what are the cards, as identified by the third section of card code only now, within this group?
                foreach (CardCodeAndCount cd in currentList)
                {
                    string code = cd.CardCode;
                    int sequenceNumber = int.Parse(code.Substring(4, 3));
                    bytes.AddRange(VarintTranslator.GetVarint(sequenceNumber));
                }
            }
        }

        public static bool ValidCardCodesAndCounts(List<CardCodeAndCount> deck)
        {
            foreach (CardCodeAndCount ccc in deck)
            {
                if (ccc.CardCode.Length != CARD_CODE_LENGTH)
                    return false;

                int parsed;
                if (!int.TryParse(ccc.CardCode.Substring(0, 2), out parsed))
                    return false;

                string faction = ccc.CardCode.Substring(2, 2);
                if (!FactionCodeToIntIdentifier.ContainsKey(faction))
                    return false;

                if (!int.TryParse(ccc.CardCode.Substring(4, 3), out parsed))
                    return false;

                if (ccc.Count < 1)
                    return false;
            }
            return true;
        }
    }
}
